@andig , testing this branch and got following error:
```
[main  ] ERROR 2024/11/24 18:09:37 creating tariff grid failed: cannot create tariff type 'entsoe': _.go:12:2: expected declaration, found charges
```

Configs:
```
tariffs:
  grid:
    type: entsoe
    domain: BZN|PT
    securitytoken: xxxx
    charges: 0
    tax: 0
    formula: (price + 0.004 + 0.0028930 + 0.010)*(1 + 0.16) + 0.005
```

If I remove formula, it works.

--- a/tariff/awattar.go
+++ b/tariff/awattar.go
@@ -86,7 +86,7 @@ func (t *Awattar) run(done chan error) {
 			ar := api.Rate{
 				Start: r.StartTimestamp.Local(),
 				End:   r.EndTimestamp.Local(),
-				Price: t.totalPrice(r.Marketprice / 1e3),
+				Price: t.totalPriceAt(r.Marketprice/1e3, r.StartTimestamp),
 			}
 			data = append(data, ar)
 		}

--- a/tariff/awattar.go
+++ b/tariff/awattar.go
@@ -86,7 +86,7 @@ func (t *Awattar) run(done chan error) {
 			ar := api.Rate{
 				Start: r.StartTimestamp.Local(),
 				End:   r.EndTimestamp.Local(),
-				Price: t.totalPriceAt(r.Marketprice/1e3, r.StartTimestamp),
+				Price: t.totalPrice(r.Marketprice/1e3, r.StartTimestamp),
 			}
 			data = append(data, ar)
 		}

--- a/tariff/edf-tempo.go
+++ b/tariff/edf-tempo.go
@@ -141,7 +141,7 @@ func (t *EdfTempo) run(done chan error) {
 				ar := api.Rate{
 					Start: ts,
 					End:   ts.Add(time.Hour),
-					Price: t.totalPriceAt(t.prices[strings.ToLower(r.Value)], ts),
+					Price: t.totalPrice(t.prices[strings.ToLower(r.Value)], ts),
 				}
 				data = append(data, ar)
 			}

--- a/tariff/edf-tempo.go
+++ b/tariff/edf-tempo.go
@@ -141,7 +141,7 @@ func (t *EdfTempo) run(done chan error) {
 				ar := api.Rate{
 					Start: ts,
 					End:   ts.Add(time.Hour),
-					Price: t.totalPrice(t.prices[strings.ToLower(r.Value)]),
+					Price: t.totalPriceAt(t.prices[strings.ToLower(r.Value)], ts),
 				}
 				data = append(data, ar)
 			}

--- a/tariff/elering.go
+++ b/tariff/elering.go
@@ -90,7 +90,7 @@ func (t *Elering) run(done chan error) {
 			ar := api.Rate{
 				Start: ts.Local(),
 				End:   ts.Add(time.Hour).Local(),
-				Price: t.totalPrice(r.Price / 1e3),
+				Price: t.totalPriceAt(r.Price/1e3, ts),
 			}
 			data = append(data, ar)
 		}

--- a/tariff/elering.go
+++ b/tariff/elering.go
@@ -85,12 +85,12 @@ func (t *Elering) run(done chan error) {
 
 		data := make(api.Rates, 0, len(res.Data[t.region]))
 		for _, r := range res.Data[t.region] {
-			ts := time.Unix(r.Timestamp, 0)
+			ts := time.Unix(r.Timestamp, 0).Local()
 
 			ar := api.Rate{
-				Start: ts.Local(),
-				End:   ts.Add(time.Hour).Local(),
-				Price: t.totalPriceAt(r.Price/1e3, ts),
+				Start: ts,
+				End:   ts.Add(time.Hour),
+				Price: t.totalPrice(r.Price/1e3, ts),
 			}
 			data = append(data, ar)
 		}

--- a/tariff/embed.go
+++ b/tariff/embed.go
@@ -82,15 +82,7 @@ func (t *embed) init() (err error) {
 	return err
 }
 
-// func (t *embed) totalPrice(price float64) float64 {
-// 	if t.calc != nil {
-// 		res, _ := t.calc(price, time.Time{})
-// 		return res
-// 	}
-// 	return (price + t.Charges) * (1 + t.Tax)
-// }
-
-func (t *embed) totalPriceAt(price float64, ts time.Time) float64 {
+func (t *embed) totalPrice(price float64, ts time.Time) float64 {
 	if t.calc != nil {
 		res, _ := t.calc(price, ts.Local())
 		return res

--- a/tariff/embed.go
+++ b/tariff/embed.go
@@ -82,13 +82,13 @@ func (t *embed) init() (err error) {
 	return err
 }
 
-func (t *embed) totalPrice(price float64) float64 {
-	if t.calc != nil {
-		res, _ := t.calc(price, time.Time{})
-		return res
-	}
-	return (price + t.Charges) * (1 + t.Tax)
-}
+// func (t *embed) totalPrice(price float64) float64 {
+// 	if t.calc != nil {
+// 		res, _ := t.calc(price, time.Time{})
+// 		return res
+// 	}
+// 	return (price + t.Charges) * (1 + t.Tax)
+// }
 
 func (t *embed) totalPriceAt(price float64, ts time.Time) float64 {
 	if t.calc != nil {

--- a/tariff/energinet.go
+++ b/tariff/energinet.go
@@ -85,11 +85,11 @@ func (t *Energinet) run(done chan error) {
 
 		data := make(api.Rates, 0, len(res.Records))
 		for _, r := range res.Records {
-			date, _ := time.Parse("2006-01-02T15:04:05", r.HourUTC)
+			ts, _ := time.Parse("2006-01-02T15:04:05", r.HourUTC)
 			ar := api.Rate{
-				Start: date.Local(),
-				End:   date.Add(time.Hour).Local(),
-				Price: t.totalPrice(r.SpotPriceDKK / 1e3),
+				Start: ts.Local(),
+				End:   ts.Add(time.Hour).Local(),
+				Price: t.totalPriceAt(r.SpotPriceDKK/1e3, ts),
 			}
 			data = append(data, ar)
 		}

--- a/tariff/energinet.go
+++ b/tariff/energinet.go
@@ -89,7 +89,7 @@ func (t *Energinet) run(done chan error) {
 			ar := api.Rate{
 				Start: ts.Local(),
 				End:   ts.Add(time.Hour).Local(),
-				Price: t.totalPriceAt(r.SpotPriceDKK/1e3, ts),
+				Price: t.totalPrice(r.SpotPriceDKK/1e3, ts),
 			}
 			data = append(data, ar)
 		}

--- a/tariff/entsoe.go
+++ b/tariff/entsoe.go
@@ -155,7 +155,7 @@ func (t *Entsoe) run(done chan error) {
 			ar := api.Rate{
 				Start: r.Start,
 				End:   r.End,
-				Price: t.totalPriceAt(r.Value, r.Start),
+				Price: t.totalPrice(r.Value, r.Start),
 			}
 			data = append(data, ar)
 		}

--- a/tariff/entsoe.go
+++ b/tariff/entsoe.go
@@ -155,7 +155,7 @@ func (t *Entsoe) run(done chan error) {
 			ar := api.Rate{
 				Start: r.Start,
 				End:   r.End,
-				Price: t.totalPrice(r.Value),
+				Price: t.totalPriceAt(r.Value, r.Start),
 			}
 			data = append(data, ar)
 		}

--- a/tariff/pun.go
+++ b/tariff/pun.go
@@ -187,19 +187,16 @@ func (t *Pun) getData(day time.Time) (api.Rates, error) {
 			return nil, fmt.Errorf("load location: %w", err)
 		}
 
-		start := time.Date(date.Year(), date.Month(), date.Day(), hour-1, 0, 0, 0, location)
-		end := start.Add(time.Hour)
-
-		priceStr := strings.Replace(p.PUN, ",", ".", -1) // Ersetzen Sie Komma durch Punkt
-		price, err := strconv.ParseFloat(priceStr, 64)
+		price, err := strconv.ParseFloat(strings.ReplaceAll(p.PUN, ",", "."), 64)
 		if err != nil {
 			return nil, fmt.Errorf("parse price: %w", err)
 		}
 
+		ts := time.Date(date.Year(), date.Month(), date.Day(), hour-1, 0, 0, 0, location)
 		ar := api.Rate{
-			Start: start,
-			End:   end,
-			Price: t.totalPrice(price / 1e3),
+			Start: ts,
+			End:   ts.Add(time.Hour),
+			Price: t.totalPriceAt(price/1e3, ts),
 		}
 		data = append(data, ar)
 	}

--- a/tariff/pun.go
+++ b/tariff/pun.go
@@ -196,7 +196,7 @@ func (t *Pun) getData(day time.Time) (api.Rates, error) {
 		ar := api.Rate{
 			Start: ts,
 			End:   ts.Add(time.Hour),
-			Price: t.totalPriceAt(price/1e3, ts),
+			Price: t.totalPrice(price/1e3, ts),
 		}
 		data = append(data, ar)
 	}

--- a/tariff/smartenergy.go
+++ b/tariff/smartenergy.go
@@ -72,7 +72,7 @@ func (t *SmartEnergy) run(done chan error) {
 			ar := api.Rate{
 				Start: r.Date.Local(),
 				End:   r.Date.Add(15 * time.Minute).Local(),
-				Price: t.totalPriceAt(r.Value/100, r.Date),
+				Price: t.totalPrice(r.Value/100, r.Date),
 			}
 			data = append(data, ar)
 		}

--- a/tariff/smartenergy.go
+++ b/tariff/smartenergy.go
@@ -72,7 +72,7 @@ func (t *SmartEnergy) run(done chan error) {
 			ar := api.Rate{
 				Start: r.Date.Local(),
 				End:   r.Date.Add(15 * time.Minute).Local(),
-				Price: t.totalPrice(r.Value / 100),
+				Price: t.totalPriceAt(r.Value/100, r.Date),
 			}
 			data = append(data, ar)
 		}

--- a/tariff/tariff.go
+++ b/tariff/tariff.go
@@ -103,7 +103,7 @@ func (t *Tariff) run(forecastG func() (string, error), done chan error) {
 				return backoff.Permanent(err)
 			}
 			for i, r := range data {
-				data[i].Price = t.totalPriceAt(r.Price, r.Start)
+				data[i].Price = t.totalPrice(r.Price, r.Start)
 			}
 			return nil
 		}, bo()); err != nil {
@@ -140,7 +140,7 @@ func (t *Tariff) priceRates() (api.Rates, error) {
 		res[i] = api.Rate{
 			Start: slot,
 			End:   slot.Add(time.Hour),
-			Price: t.totalPriceAt(price, slot),
+			Price: t.totalPrice(price, slot),
 		}
 	}
 

--- a/tariff/tibber.go
+++ b/tariff/tibber.go
@@ -117,7 +117,7 @@ func (t *Tibber) rates(pi []tibber.Price) api.Rates {
 	for _, r := range pi {
 		price := r.Total
 		if t.Charges != 0 || t.Tax != 0 {
-			price = t.totalPrice(r.Energy)
+			price = t.totalPriceAt(r.Energy, r.StartsAt)
 		}
 		ar := api.Rate{
 			Start: r.StartsAt.Local(),

--- a/tariff/tibber.go
+++ b/tariff/tibber.go
@@ -117,7 +117,7 @@ func (t *Tibber) rates(pi []tibber.Price) api.Rates {
 	for _, r := range pi {
 		price := r.Total
 		if t.Charges != 0 || t.Tax != 0 {
-			price = t.totalPriceAt(r.Energy, r.StartsAt)
+			price = t.totalPrice(r.Energy, r.StartsAt)
 		}
 		ar := api.Rate{
 			Start: r.StartsAt.Local(),

